pkg/snmp/traps/listener.go

```diff
@@ -0,0 +1,85 @@
+package traps
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/log"
+	"github.com/gosnmp/gosnmp"
+)
+
+// TrapListener opens an UDP socket and put all received traps in a channel
+type TrapListener struct {
+	config        Config
+	packets       PacketsChannel
+	listener      *gosnmp.TrapListener
+	errorsChannel chan error
+	errorLogger   *log.ThrottledLogger
+}
+
+// NewTrapListener creates a simple TrapListener instance but does not start it
+func NewTrapListener(config Config, packets PacketsChannel) (*TrapListener, error) {
+	var err error
+	gosnmpListener := gosnmp.NewTrapListener()
+	gosnmpListener.Params, err = config.BuildSNMPParams()
+	if err != nil {
+		return nil, err
+	}
+	errorsChan := make(chan error, 1)
+	trapListener := &TrapListener{
+		config:        config,
+		packets:       packets,
+		listener:      gosnmpListener,
+		errorsChannel: errorsChan,
+		errorLogger:   log.NewThrottled(5, 10*time.Second),
+	}
+
+	gosnmpListener.OnNewTrap = trapListener.receiveTrap
+	return trapListener, nil
+}
+
+// Start the TrapListener instance. Need to be manually Stopped
+func (t *TrapListener) Start() error {
+	log.Infof("Start listening for traps on %s", t.config.Addr())
+	go t.run()
+	return t.blockUntilReady()
+}
+
+func (t *TrapListener) run() {
+	err := t.listener.Listen(t.config.Addr()) // blocking call
+	if err != nil {
+		t.errorsChannel <- err
+	}
+
+}
+
+func (t *TrapListener) blockUntilReady() error {
+	select {
+	// Wait for listener to be started and listening to traps.
+	// See: https://godoc.org/github.com/gosnmp/gosnmp#TrapListener.Listening
+	case <-t.listener.Listening():
+		return nil
+	// If the listener failed to start (eg because it couldn't bind to a socket),
+	// we'll get an error here.
+	case err := <-t.errorsChannel:
+		return fmt.Errorf("error happened when listening for SNMP Traps: %s", err)
+	}
+}
+
+// Stop the current TrapListener instance
+func (t *TrapListener) Stop() {
+	t.listener.Close()
+}
+
+func (t *TrapListener) receiveTrap(p *gosnmp.SnmpPacket, u *net.UDPAddr) {
+	currentTime := time.Now().UnixMilli()
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
No need to init currentTime in case of error, you could replace this after the error check